pdns/recursor_cache.hh

```diff
@@ -201,55 +201,60 @@ private:
   > ecsIndex_t;
 
   typedef std::pair<NameAndRTagOnlyHashedTagIterator_t, NameAndRTagOnlyHashedTagIterator_t> Entries;
-  
+
   struct MapCombo
   {
     MapCombo() {}
     MapCombo(const MapCombo &) = delete;
     MapCombo & operator=(const MapCombo &) = delete;
-    cache_t d_map;
-    ecsIndex_t d_ecsIndex;
-    DNSName d_cachedqname;
-    OptTag d_cachedrtag;
-    Entries d_cachecache;
-    bool d_cachecachevalid{false};
-    std::atomic<uint64_t> d_entriesCount{0};
-    uint64_t d_contended_count{0};
-    uint64_t d_acquired_count{0};
-
-    void invalidate()
+    struct LockedContent
     {
-      d_cachecachevalid = false;
-    }
+      cache_t d_map;
+      ecsIndex_t d_ecsIndex;
+      DNSName d_cachedqname;
+      OptTag d_cachedrtag;
+      Entries d_cachecache;
+      uint64_t d_contended_count{0};
+      uint64_t d_acquired_count{0};
+      bool d_cachecachevalid{false};
+
+      void invalidate()
+      {
+        d_cachecachevalid = false;
+      }
+    };
+
+    LockGuarded<LockedContent> d_content;
+    std::atomic<uint64_t> d_entriesCount{0};
   };
 
-  vector<LockGuarded<MapCombo>> d_maps;
-  LockGuarded<MapCombo>& getMap(const DNSName &qname)
+  vector<MapCombo> d_maps;
+  MapCombo& getMap(const DNSName &qname)
   {
     return d_maps.at(qname.hash() % d_maps.size());
   }
 
   static time_t fakeTTD(OrderedTagIterator_t& entry, const DNSName& qname, QType qtype, time_t ret, time_t now, uint32_t origTTL, bool refresh);
 
   bool entryMatches(OrderedTagIterator_t& entry, QType qt, bool requireAuth, const ComboAddress& who);
-  Entries getEntries(MapCombo& map, const DNSName &qname, const QType qt, const OptTag& rtag);
-  cache_t::const_iterator getEntryUsingECSIndex(MapCombo& map, time_t now, const DNSName &qname, QType qtype, bool requireAuth, const ComboAddress& who);
+  Entries getEntries(MapCombo::LockedContent& content, const DNSName &qname, const QType qt, const OptTag& rtag);
+  cache_t::const_iterator getEntryUsingECSIndex(MapCombo::LockedContent& content, time_t now, const DNSName &qname, QType qtype, bool requireAuth, const ComboAddress& who);
 
-  time_t handleHit(MapCombo& map, OrderedTagIterator_t& entry, const DNSName& qname, uint32_t& origTTL, vector<DNSRecord>* res, vector<std::shared_ptr<RRSIGRecordContent>>* signatures, std::vector<std::shared_ptr<DNSRecord>>* authorityRecs, bool* variable, boost::optional<vState>& state, bool* wasAuth, DNSName* authZone);
+  time_t handleHit(MapCombo::LockedContent& content, OrderedTagIterator_t& entry, const DNSName& qname, uint32_t& origTTL, vector<DNSRecord>* res, vector<std::shared_ptr<RRSIGRecordContent>>* signatures, std::vector<std::shared_ptr<DNSRecord>>* authorityRecs, bool* variable, boost::optional<vState>& state, bool* wasAuth, DNSName* authZone);
 
 public:
-  static LockGuardedTryHolder<MapCombo> lock(LockGuarded<MapCombo>& map)
+  static LockGuardedTryHolder<MapCombo::LockedContent> lock(LockGuarded<MapCombo::LockedContent>& content)
```

Comment:
I think it would be nicer if `lock` would be a method of `LockedContent` instead of a `static` method of `MemRecursorCache`.

Same for negcache. I have a local version with those changed here if you agree.

Reply:
Good idea!

Reply:
Committed